### DISCUSSION TITLE
Synchronisation en temps réel des missions

### DISCUSSION THREAD

**Discussion by louptheron** [MTES-MCT/monitorfish#2573] [Ideas]
## Besoins

Nous avons besoin d'avoir des missions synchronisées presque en temps réel, voici les deux cas :
- Une mission enregistrée au CACEM doit se voir instantanément au CNSP (et vice-versa)
- Une mission dont le formulaire est ouvert en édition au CACEM doit rester synchronisée avec toute modification faite au CNSP (et vice-versa) ou dans son centre

## Process utilisateur

On peut **1) garder le même principe qu'actuellement (l'utilisateur sauvegarde une mission au clic)** ou 2) on **"auto-save" le formulaire à chaque modification**.

Pour le **cas 1)**, étant donné qu'il peut y avoir des modifications non-repercutées, il faut définir un process de mise-à-jour lorsqu'un formulaire ouvert est modifié sur un autre poste, voici un exemple de process.

Donc, lorsque le formulaire des missions est ouvert est qu'une mission est modifiée sur un autre poste :
  - Si la partie gauche du formulaire est modifiée (ex. l'unité est modifiée), on affiche un bandeau sur cette partie gauche pour demander à l'utilisateur de mettre à jour la mission
  - Si la partie droite du formulaire est modifiée
    - Si la modification provient du même centre (étant au CACEM, une action CACEM est modifiée sur Env), on affiche un bandeau pour demander à l'utilisateur de mettre à jour la mission
    - Si la modification provient de l'autre centre (étant au CNSP, une action CACEM est modifiée sur Env), on peut mettre à jour automatiquement les actions car il n'y a pas d'overlap entre les deux centres

Pour le **cas 2)**, étant donné que tout modification est repercutée en temps-réel sur les autres formulaires, on peut se dire qu'un process de mise-à-jour n'est pas requis (pas besoin de prévenir l'utilisateur car la portée de la modification est minime) et qu'une modification peut instantanément mettre à jour un formulaire sans "demander la permission".

Conditions d'envoi du formulaire lorsqu'un champ est modifié :
- Si la modification d'un champ invalide le formulaire (ex. champ unité manquant), alors la mission n'est pas "auto-save" et un message d'erreur classique s'affiche en-dessous du champ
- Lors de la création d'une mission, le 1er "auto-save" est envoyé lorsque le 1er champ est rempli ; les autres champs seront affichés en rouge

Donc, lorsque le formulaire des missions est ouvert est qu'une mission est modifiée sur un autre poste :
- La nouvelle version de la mission récupérée met à jour le formulaire ouvert (par contre, si deux utilisateurs écrivent en même temps, un utilisateur peut perdre son travail).

> On peut trancher avec ces questions :
> - Côté utilisateur, quel process convient le mieux ?
> - Côté dev, quel process est le plus facile à implémenter ?
    
## Implémentation

Il y a deux sujets techniques à couvrir : **1) comment recevoir les modifications** et 2) **comment enregistrer les modifications**.

### 1) Réception des modifications

Aujourd'hui la synchronisation se fait par `polling` régulier toutes les 2 minutes.

Je propose de faire un Server-Side Event - SSE - (un flux continu d'évenement unidirectionnel du backend) pour recevoir un évenement de mise-à-jour en temps réel sur les missions car :
- C'est bien pris en charge par notre framework Spring
- C'est déjà géré par le navigateur ([EventSource](https://developer.mozilla.org/fr/docs/Web/API/EventSource))

Nous pouvons utiliser ce canal réactif seulement pour invalider le cache `RTK` et donc demander une nouvelle version (sans la mission sans le `payload`) ou insérer directement la nouvelle mission (reçue dans le `payload`) dans `RTK`.

Gestion du cache RTK avec des événements : https://redux-toolkit.js.org/rtk-query/usage/streaming-updates#websocket-chat-api

### 2) Enregistrement des modifications

Pour moi, le process 2) auto-save paraît le plus simple à mettre en oeuvre : toute modification du formulaire `Formik` (avec potentiellement un petit `debounce`) entraine la mise-à-jour de la mission via l'API (pas besoin de canal de type WebSocket dans ce cas à mon sens).

Cet enregistrement envoie depuis le backend un événement SSE `MISSION_UPDATE` à tous les `listeners` (avec ou sans la mission en `payload` - à définir).

> Etant donné que les missions s'affichent et se synchronisent sur les autres postes, elles doivent être validées par `Yup` lors de l'enregistrement.

----

Qu'en pensez-vous ?

**Comment by AdelineCelier** [comment]
@louptheron je me rends compte qu'on n'a pas reparlé de la manière dont seraient gérées les validations de champ dans le cas d'un auto-save. Est-ce que tu pourrais le préciser dans cette discussion afin qu'on ait toutes les données du problème ? merci !

**Reply by louptheron** [reply]
Ok j'ai rajouté cette partie dans le message initial.
Pour moi on valide le formulaire à chaque modification et on affiche un message d'erreur classique sous le champ si ça ne passe pas la validation.
